src/components/Asset/AssetActions/Pool/Add/index.tsx

```diff
@@ -80,38 +81,58 @@ export default function Add({
 
   // Get maximum amount for either OCEAN or datatoken
   useEffect(() => {
-    if (!accountId || !isAssetNetwork || !poolAddress) return
+    if (!web3 || !accountId || !isAssetNetwork || !poolAddress) return
 
     async function getMaximum() {
-      // const amountMaxPool =
-      //   coin === 'OCEAN'
-      //     ? await ocean.pool.getOceanMaxAddLiquidity(poolAddress)
-      //     : await ocean.pool.getDTMaxAddLiquidity(poolAddress)
-      // const amountMax =
-      //   coin === 'OCEAN'
-      //     ? Number(balance.ocean) > Number(amountMaxPool)
-      //       ? amountMaxPool
-      //       : balance.ocean
-      //     : Number(dtBalance) > Number(amountMaxPool)
-      //     ? amountMaxPool
-      //     : dtBalance
-      // setAmountMax(Number(amountMax).toFixed(3))
+      const poolInstance = new Pool(web3, LoggerInstance)
+      const baseTokenAddress = await poolInstance.getBasetoken(poolAddress)
+      const tokenInAddress = coin === 'OCEAN' ? baseTokenAddress : dtAddress
+      setTokenInAddress(tokenInAddress)
+
+      const amountMaxPool = await poolInstance.getReserve(
+        poolAddress,
+        tokenInAddress
+      )
+
+      // coin === 'OCEAN'
+      //   ? await poolInstance.getOceanMaxAddLiquidity(poolAddress)
+      //   : await poolInstance.getDTMaxAddLiquidity(poolAddress)
+      const amountMax =
+        coin === 'OCEAN'
+          ? Number(balance.ocean) > Number(amountMaxPool)
+            ? amountMaxPool
+            : balance.ocean
+          : Number(dtBalance) > Number(amountMaxPool)
+          ? amountMaxPool
+          : dtBalance
+      setAmountMax(Number(amountMax).toFixed(3))
     }
     getMaximum()
-  }, [accountId, isAssetNetwork, poolAddress, coin, dtBalance, balance.ocean])
+  }, [
+    web3,
+    accountId,
+    isAssetNetwork,
+    poolAddress,
+    dtAddress,
+    coin,
+    dtBalance,
+    balance.ocean
+  ])
 
   // Submit
   async function handleAddLiquidity(amount: number, resetForm: () => void) {
+    const poolInstance = new Pool(web3, LoggerInstance)
+    const minPoolAmountOut = '0' // ? TODO: how to get?
```

Comment:
how to calculate this for `joinswapExternAmountIn()`? Doe we have ocean.js method or contract method for this or is this supposed to be calculated based on user input value combined with slippage value?

Reply:
this needs some thinking, i think fees should be taken in consideration as well

Reply:
good point, moved for now to https://github.com/oceanprotocol/market/issues/1027